graph-builder/src/metadata.rs

```diff
@@ -0,0 +1,95 @@
+//! This module implements the fetching of dynamic metadata
+
+use failure::{Error, Fallible};
+use futures::future::Future;
+use quay;
+use registry;
+use std::path::PathBuf;
+
+pub static DEFAULT_QUAY_LABEL_FILTER: &str = "com.openshift.upgrades.graph";
+pub static MANIFESTREF_KEY: &str = "com.openshift.upgrades.graph.release.manifestref";
+
+pub trait AsyncMetadataFetcher<T, U, E> {
+    fn fetch_metadata(&self, fetch_args: U) -> Box<Future<Item = T, Error = E>>;
+}
+
+/// Convenience type for a metadata entry
+type MetadataEntry = (String, String);
+
+/// Convenience type for fetch_args()'s metadata_fetcher argument
+pub type MetadataFetcher = Box<AsyncMetadataFetcher<Vec<MetadataEntry>, String, Error>>;
+
+pub struct QuayMetadataFetcher {
+    client: quay::v1::Client,
+    label_filter: String,
+    repo: String,
+}
+impl AsyncMetadataFetcher<Vec<MetadataEntry>, String, Error> for QuayMetadataFetcher {
+    fn fetch_metadata(
+        &self,
+        manifestref: String,
+    ) -> Box<Future<Item = Vec<MetadataEntry>, Error = Error>> {
+        Box::new(
+            self.client
+                .get_labels(
+                    self.repo.clone(),
+                    manifestref,
+                    Some(self.label_filter.clone()),
+                )
+                .map(|labels| labels.into_iter().map(Into::into).collect()),
+        )
+    }
+}
+
+impl QuayMetadataFetcher {
+    pub fn try_new(
+        label_filter: String,
+        api_token_path: Option<&PathBuf>,
+        api_base: String,
+        repo: String,
+    ) -> Fallible<Box<Self>> {
+        let api_token =
+            quay::read_credentials(api_token_path).expect("could not read quay API credentials");
+
+        let client: quay::v1::Client = quay::v1::Client::builder()
+            .access_token(api_token.map(|s| s.to_string()))
+            .api_base(Some(api_base.to_string()))
+            .build()?;
+
+        Ok(Box::new(QuayMetadataFetcher {
+            client,
+            label_filter,
+            repo,
+        }))
+    }
+}
+
+/// Asynchronously fetches and populates the dynamic metadata for the given releases
+pub fn fetch_and_populate_dynamic_metadata<'a>(
+    metadata_fetcher: &'a MetadataFetcher,
+    releases: Vec<registry::Release>,
+) -> impl Future<Item = Vec<registry::Release>, Error = Error> + 'a {
+    let populated_releases = releases.into_iter().map(move |mut release| {
+        futures::future::ok(release.metadata.metadata.remove(MANIFESTREF_KEY))
+            .and_then(|manifestref_value| match manifestref_value {
+                Some(manifestref) => Ok(manifestref),
+                None => bail!("release metadata doesn't contain the manifestref"),
```

Comment:
Should we maybe just do this filtering upfront (outside of the future) and iterate on a vector of `(release, ref)`? I'm worried that a single entry without the corresponding manifest-ref will fail the whole `join_all`, while we could instead just skip the specific one.

Reply:
This was intentional and I'm happy to discuss it. I assume that we have an incomplete view on the graph if we can't get *all* metadata, so I'd rather report a failure than a graph which was not intended to be published. Similarly it will error out if a single fetching operation fails.

Reply:
I'm fine with this intentional behavior then. I would only suggest to make that more explicit (and actionable) to both human operators and casual code readers by:
 * having a docstring to mention the "either all or none" failure mode
 * adding the version id of the faulty release to the bail message

Reply:
Ack, I do appreciate that docstring, thanks!